### PR TITLE
Regression Tests(addition to the PR: try compiling mfem with gitlab actions)

### PR DESCRIPTION
**Pull request by kevinhkhuynh**
The goal here is to compile libROM with MFEM for Gitlab actions so that we can run the example scripts as part of a test suite.

### CODE REVIEW

**Comment by chldkdtn** [comment]
@kevinhkhuynh do you think you need to pass this to @pranav-vempati as well?

File: .github/workflows/ci.yml
Comment (kevinhkhuynh) [inline .github/workflows/ci.yml:22]: Here, I'm trying to get the Github workflows to compile with the -m option, which builds MFEM as well, but it isn't working... We need this to be able to run the examples we want in the test suite.

File: .github/workflows/ci.yml
Comment (kevinhkhuynh) [inline .github/workflows/ci.yml:25]: I just tried this to see if it would work. This doesn't have to be macos-12. It could go back to macos-latest.

File: .github/workflows/run_tests/action.yml
Comment (kevinhkhuynh) [inline .github/workflows/run_tests/action.yml:7]: This line is executing one of the examples. It doesn't make it to here because it crashes a bit earlier in the build process.

**Comment by kevinhkhuynh** [comment]
For some reason, the build process is hanging/crashing during the MFEM build process.

**Comment by kevinhkhuynh** [comment]
> For some reason, the build process is hanging/crashing during the MFEM build process for the MAC. You can see this when you look into the "Details" in the CI tests below. For Linux, the test is failing at the example run due to "Operating system error: Cannot allocate memory"

@axla-io Axel, didn't you run into a similar issue on your virtual machine? How did you end up fixing it?

File: .github/workflows/run_tests/action.yml
Comment (kevinhkhuynh) [inline .github/workflows/run_tests/action.yml:26]: This line is executing one of the examples. It is crashing here on Linux, due to an memory allocation issue. Mac is running fine. This may require updating some settings on Github Actions to increase the memory allocation or changing some build settings in libROM.

**Comment by axla-io** [comment]
> > For some reason, for Linux, the test is failing at the example run due to "Operating system error: Cannot allocate memory"
> 
> 
> 
> @axla-io Axel, didn't you run into a similar issue on your virtual machine, relating to the cannot allocate memory? How did you end up fixing it?

Yes I did! In the end I allocated more memory and more swap memory for my vm. Did you set the _mc-model = large_ flag also?

File: .github/workflows/run_tests/action.yml
Comment (axla-io) [reply .github/workflows/run_tests/action.yml]: Hi @kevinhkhuynh @pranav-vempati sorry for being away. Is this working now? The way I solved it on my machine was to increase memory allocation. The settings I have is:

memory=8GB
swap memory=7GB

File: regression_tests/basisComparator.cpp
Comment (dylan-copeland) [inline regression_tests/basisComparator.cpp:1]: I suppose the copyright heading should be added to the cpp files (multiple new files in this PR).

**Comment by dylan-copeland** [comment]
@pranav-vempati Should `regression_tests/simulationLogFile` be in this PR?

**Comment by pranav-vempati** [comment]
All suggestions except the comparison of examples' relative errors are now incorporated in this PR.

**Comment by chldkdtn** [comment]
@dylan-copeland @siuwuncheung , @pranav-vempati has just informed me that this PR is ready for review except continuous integration part. All the regression tests must be working locally, so we can start reviewing the local regression test implementation. I suggest not to wait until the CI is fixed. Let's start giving @pranav-vempati valuable feedbacks on this PR.

**Comment by dylan-copeland** [comment]
@pranav-vempati I ran the regression tests and got this result in `regression_tests/results/de_parametric_heat_conduction_greedy.log`:
`More processors requested than permitted`
I'm not sure how many processors are requested, but the examples in libROM should not need more than 4 processors.

File: regression_tests/run_regression_tests.sh
Comment (dylan-copeland) [inline regression_tests/run_regression_tests.sh:9]:
```suggestion
# sbatch ./regression_tests/run_regression_tests.sh
```

**Review by dylan-copeland — APPROVED** [review]
Thanks @pranav-vempati! The tests all pass on linux and mac. One small suggestion is that the squash-and-merge has resulted in some very long commit messages in the libROM history. Since this PR has almost 200 commits, it would be a good idea to replace all those commit messages with a brief summary of this PR.

**Comment by chldkdtn** [comment]
@pranav-vempati, I intentionally broke one of dmd examples and ran regression tests. The regression tests correctly identified the failing example, but I cannot find any log or output files which tells me what exactly broke. If there is such an output file, can you tell me where? If not, can you modify the regression test so that it outputs the error or the reason for failure?

File: regression_tests/REGRESSIONTEST.md
Comment (siuwuncheung) [inline regression_tests/REGRESSIONTEST.md:29]: It is better to not append the comment after the command. Same on Line 31

**Review by siuwuncheung — CHANGES_REQUESTED** [review]

**Comment by chldkdtn** [comment]
@pranav-vempati,  can you add instruction of how to add a new regression test?

File: .github/workflows/run_tests/action.yml
Comment (pranav-vempati) [reply .github/workflows/run_tests/action.yml]: Fixed.

File: CMakeLists.txt
Comment (pranav-vempati) [reply CMakeLists.txt]: Fixed.

File: regression_tests/REGRESSIONTEST.md
Comment (pranav-vempati) [reply regression_tests/REGRESSIONTEST.md]: Fixed.

**Comment by pranav-vempati** [comment]
@chldkdtn Instructions on how to add a new regression test have been added to REGRESSIONTEST.md.

File: regression_tests/REGRESSIONTEST.md
Comment (pranav-vempati) [reply regression_tests/REGRESSIONTEST.md]: Fixed.

File: regression_tests/run_regression_tests.sh
Comment (pranav-vempati) [reply regression_tests/run_regression_tests.sh]: > Thanks @pranav-vempati! The tests all pass on linux and mac. One small suggestion is that the squash-and-merge has resulted in some very long commit messages in the libROM history. Since this PR has almost 200 commits, it would be a good idea to replace all those commit messages with a brief summary of this PR.

 @dylan-copeland, thanks for the suggestion. I will do this before I merge this PR.

**Comment by pranav-vempati** [comment]
> Thanks @pranav-vempati! The tests all pass on linux and mac. One small suggestion is that the squash-and-merge has resulted in some very long commit messages in the libROM history. Since this PR has almost 200 commits, it would be a good idea to replace all those commit messages with a brief summary of this PR.

@dylan-copeland Thanks for the suggestion. I will do this before I merge this PR.

**Review by siuwuncheung — APPROVED** [review]

**Review by chldkdtn — APPROVED** [review]

**Pull request merged**
